### PR TITLE
Feature/issue 623 add wa sv2 for export/reports download

### PR DESCRIPTION
**Pull request by fullofentropy**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #623

### Added

- Added support for v2 Web Application Scanning for:
  - get scan details https://developer.tenable.com/reference/was-v2-scans-details
  - export scan results https://developer.tenable.com/reference/was-v2-scans-export
  - download exported scans https://developer.tenable.com/reference/was-v2-scans-download-export

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test Code used attached
- [x] Executed Regression tests

**Test Configuration**: Kali Linux VM with cloud.tenable.com APIs
* Python Version(s) 3.8:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by fullofentropy** [comment]
My apologies, I clicked the button too soon to start a pull request, I have done my own unit testing but haven't figured out regression tests for this project.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,14 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.4.8]
```
Comment (aseemsavio) [inline CHANGELOG.md:7]: Please remove the changes made to this file. pyTenable - version 1.4.8 is already out. Instead, you could add these information to the PR description.

File: tenable/io/scans.py [hunk 2]
```diff
@@ -849,6 +849,13 @@ def results(self, scan_id, history_id=None, history_uuid=None):
             params['history_uuid'] = self._check(
                 'history_uuid', history_uuid, 'scanner-uuid')
 
+        if 'web-app2' in kw.values():
```
Comment (aseemsavio) [inline tenable/io/scans.py:852]: Please document the new kew word argument - `web-app2` for the `readthedocs` doc generation.

You could take the following as your example:

```
scan_type (str, optional):
                This parameter is required only when using the API with
                Web Application Scanning. Available option is 'web-app'.
```

File: tenable/io/scans.py [hunk 3]
```diff
@@ -981,36 +989,74 @@ def f(response: requests.Response,
...
-            # https://github.com/tenable/pyTenable/issues/305
-            stream_hook(resp, fobj, chunk_size=1024)
+        # Inserting was v2 API implimentation conditional for exporting/downloading reports
+        if 'web-app2' in kw.values():
+            if kw['format'] in ['csv','html','xml']:
+                requested_file = f"text/{kw['format']}"
+            elif kw['format'] in ['json','pdf']:
+                requested_file = f"application/{kw['format']}"
+            else:
+                self._api._log.info("improper or no format passed in. defaulting to json")
```
Comment (aseemsavio) [inline tenable/io/scans.py:999]: Consider capitalising the first word in each sentence.

File: tenable/io/scans.py [hunk 4]
```diff
@@ -981,36 +989,74 @@ def f(response: requests.Response,
...
+                requested_file="application/json"
+            headers = {
+                "accept":"application/json",
+                "Content-Type":requested_file
+            }
+            # put request to start generation of the request report
+            # Response: 200-completed; 202-request accepted
+            fid = self._api.put('was/v2/scans/{}/report'.format(scan_id),
+                                headers=headers)#.json()['file']
```
Comment (aseemsavio) [inline tenable/io/scans.py:1008]: Can the commented out piece of code be removed?

File: tenable/io/scans.py [hunk 5]
```diff
@@ -981,36 +989,74 @@ def f(response: requests.Response,
...
+            if str(fid).find('20') < 0:
+                self._api._log.error(f"There was an problem requesting report:{requested_file}")
+                return None
+
+            self._api._log.debug('Initiated scan export {}'.format(requested_file))
+            # This is basically a do while loop in python used to retrieve response object.
+            while True:
+                resp = self._api.get('was/v2/scans/{}/report'.format(scan_id),
+                                    headers=headers, stream=True)#.json()['file']
```
Comment (aseemsavio) [inline tenable/io/scans.py:1017]: Please check if the previous comment applies here too.

**Review by aseemsavio — CHANGES_REQUESTED** [review]
Hello @fullofentropy, Thanks for the PR. I've added a few comments. In addition, please add tests confirming the behaviour.

**Comment by aseemsavio** [comment]
Hey @fullofentropy! pyTenable majorly uses [pytest](https://docs.pytest.org/en/7.1.x/) and [pytest-vcr](https://pypi.org/project/pytest-vcr/). Looking these libraries up would be beneficial. `pyTenable/tests/io/conftest.py` houses the test fixtures. You can refer to the files inside `pyTenable/tests/io/*` directory with a `test_` prefix as your examples for tests. Thanks.

**Comment by aseemsavio** [comment]
Hi! If this PR fixes #622, please ensure it's properly written in the PR description (without space between "#" and 622). Thanks!

File: CHANGELOG.md (hunk 1 above)
Comment (fullofentropy) [reply CHANGELOG.md]: Will do, thanks

**Comment by fullofentropy** [comment]
@aseemsavio, I will add some pytests to confirm the behavior, question below as well.

> Hello @fullofentropy, Thanks for the PR. I've added a few comments. In addition, please add tests confirming the behaviour.

- I will make updates directly to  https://github.com/tenable/pyTenable/blob/master/tests/io/test_scans.py 
- After I add my unit tests, when running pytest.   Is completion of testing.yml when executed as a pull request into my forked master (which is synced with its parent)

image

**Comment by fullofentropy** [comment]
So in the current feature branch, I have added 3 pytest unit tests that currently fail because I do not know of a WAS scan_id or history_id I can use for testing.  Could someone assist me with this?  I also installed the html report (in the zip) generator for the pytest execution on this branch.  

The regression test was ran and passed using the fork, it is my added (and incomplete) tests that are failing.
[pytest_report_issue_report.zip](https://github.com/tenable/pyTenable/files/9940646/pytest_report_issue_report.zip)

_Note_: This is the first time I am using pytest, so I spent some time learning it and some of the extended features this project uses.

**Comment by fullofentropy** [comment]
FYI: I did test the implementation using my own tenable credentials and WAS based scans.

**Comment by aseemsavio** [comment]
Hi @fullofentropy. I'll try and explain as best I can.

You must have seen your test function accepting `api` as an argument. The [pytest fixture](https://github.com/tenable/pyTenable/blob/74e0ba7c49e9447db8e5d1e3d8f567d5ba8725c8/tests/io/conftest.py#L24) named `api` injects a TenableIO object with the credentials mentioned into all the test functions that accepts a param named `api`. 

As you may be familiar, pytest records the HTTP request and response in the cassette (yaml) files, so that it can use the recorded response during the subsequent test runs instead of actually making HTTP calls every time. 

To achieve this, you need to put valid credentials in the `api` [fixture](https://github.com/tenable/pyTenable/blob/74e0ba7c49e9447db8e5d1e3d8f567d5ba8725c8/tests/io/conftest.py#L24) function and run it once (locally before committing the code). This will call the APIs with your credentials; record the HTTP responses; and creates cassette files. The next time you run the test, the APIs won't be called. Instead, the response will be fetched from the cassette files. Note that this won't store your credentials in the cassette files. 

You must make sure to remove your credential from the `api` [fixture](https://github.com/tenable/pyTenable/blob/74e0ba7c49e9447db8e5d1e3d8f567d5ba8725c8/tests/io/conftest.py#L24) before you commit your changes. You need to also commit the generated cassette files.

Hope this helps. Let me know if you need any more information.

**Pull request closed without merging**